test/distributed/pipelining/test_backward.py

```diff
@@ -53,6 +57,127 @@ def test_stage_backward(self):
                 print(f"Gradient test failed for {name}: {p.grad} vs {ref_p.grad}")
                 raise
 
+    def test_stage_backward_input(self):
+        # MLP as a stage module
+        mod = MLPModule(d_hid)
+        x = torch.randn(batch_size, d_hid)
+        # As in a pipeline stage, the inputs to this stage requires gradients
+        x.requires_grad_(True)
+        target = torch.randn(batch_size, d_hid)
+        loss_fn = torch.nn.MSELoss(reduction="sum")
+
+        # Make a copy
+        ref_mod = copy.deepcopy(mod)
+        ref_x = x.detach().requires_grad_(x.requires_grad)
+        ref_target = target.detach()
+
+        # Forward, then backward of loss with respect to inputs
+        out = mod(x)
+        loss = loss_fn(out, target)
+        dinputs, param_groups = stage_backward_input(
+            stage_outputs=(loss,),
+            output_grads=None,
+            stage_inputs=[x],
+            weights=mod.parameters(),
+        )
+
+        # Run reference
+        ref_out = ref_mod(ref_x)
+        ref_loss = loss_fn(ref_out, ref_target)
+        ref_loss.backward()
+
+        torch.testing.assert_close(x.grad, ref_x.grad)
+
+    def test_stage_backward_weight(self):
+        # MLP as a stage module
+        mod = MLPModule(d_hid)
+        x = torch.randn(batch_size, d_hid)
+        # As in a pipeline stage, the inputs to this stage requires gradients
+        x.requires_grad_(True)
+        target = torch.randn(batch_size, d_hid)
+        loss_fn = torch.nn.MSELoss(reduction="sum")
+
+        # Make a copy
+        ref_mod = copy.deepcopy(mod)
+        ref_x = x.detach().requires_grad_(x.requires_grad)
+        ref_target = target.detach()
+
+        # Forward, then backward of loss with respect to inputs
+        out = mod(x)
+        loss = loss_fn(out, target)
+        dinputs, param_groups = stage_backward_input(
+            stage_outputs=(loss,),
+            output_grads=None,
+            stage_inputs=[x],
+            weights=mod.parameters(),
+        )
+
+        # backward of loss with respect to weights
+        dweights = stage_backward_weight(mod.parameters(), param_groups)
```

Comment:
jw, is there a requirement that users call backward_input before backward_weight?

or if they just call backward_weight, will that work too?  (and if they subsequently call backward_input, would that reuse the intermediates from backward_weight?)

This doesn't really matter (i dont expect people to want to call in the wrong order) but might be a factor for 'api hardening'

Reply:
To call backward_weight the user would need the param groups which is outputted from backward_input. Theoretically they could get the parameter groups separately and just call backward_weight on its own.

> if they just call backward_weight, will that work too? (and if they subsequently call backward_input, would that reuse the intermediates from backward_weight?)

Currently no, we dont have support in backward_input to use the intermediates.

----- 
For the framework itself, I updated the error messaging so that if they are using pipeline stages (e.g. building their own schedules), then they will get an assert error: 

"Attempted to run backward_weight_one_chunk for chunk {bwd_chunk_id} without first calling `backward_one_chunk(full_backward=False)`"

and same for if they are using the dw_builder()